# Networks and Paths

## Networks

The image below shows an example of  a *network*. The yellow circles are network **nodes**. Arrows are network **edges**, and they indicate which node is connected to which other nodes. A double headed arrow 
${\tt i \leftrightarrow j}$ means that there is an edge ${\tt i \rightarrow j}$ and an edge ${\tt i\leftarrow j}$.

<br/>
<img src="https://raw.githubusercontent.com/bbadzioch/MTH309_F2019/master/notebooks_2024/network.png" style="width: 300px;">

<br/>


Networks are used to represent a variety of structures. For example:

- **Computer networks:** e.g. which computer can send data to which other computers.
- **Networks of webpages:** which webpage has links to which other webpages.
- **Transportation networks:** e.g. which airport has a direct connection to which other airports.
- **Social networks:** e.g. who likes whom.



## Adjacency matrix

An **adjacency matrix** is a  way of representing a network mathematically. It is a square matrix $A$ whose rows and columns are labeled by nodes of the network. If $a_{\tt ij}$ denotes the entry of $A$ in the column $\tt j$ and row $\tt i$, then 

<br/>

$$
a_{\tt ij} = 
\begin{cases}
1 & \text{if the network has an edge  $\tt{i {\color{red}\leftarrow} j}$} \\
0 & \text{otherwise} \\
\end{cases}
$$


For example, the adjacency matrix of the network shown above looks as follows:

<br/>

$$
\begin{matrix}
\phantom{0} & \phantom{\leftarrow}\\
\end{matrix}
\hskip 3.5mm
\begin{matrix}
{\tt \color{blue}1} & {\tt \color{blue}2} & {\tt \color{blue}3} & {\tt \color{blue}4} & {\tt \color{blue}5} \\
{\tt \color{blue}\downarrow} & {\tt \color{blue}\downarrow} & {\tt \color{blue}\downarrow} & {\tt \color{blue}\downarrow} & {\tt \color{blue}\downarrow} \\
\end{matrix}
$$

$$
\begin{matrix}
{\tt \color{blue}1} & {\tt \color{blue}\leftarrow}\\ 
{\tt \color{blue}2} & {\tt \color{blue}\leftarrow}\\ 
{\tt \color{blue}3} & {\tt \color{blue}\leftarrow}\\ 
{\tt \color{blue}4} & {\tt \color{blue}\leftarrow}\\ 
{\tt \color{blue}5} & {\tt \color{blue}\leftarrow}\\ 
\end{matrix}
\hskip 3mm
\begin{bmatrix}
0 & 1 & 0 & 1 & 0 \\
0 & 0 & 0 & 1 & 0 \\
1 & 0 & 0 & 1 & 1 \\
0 & 1 & 1 & 0 & 1 \\
0 & 1 & 0 & 1 & 0 \\
\end{bmatrix}
$$

## Network paths

A **path** in a network is a sequence of edges, such that each edge ends at a node at which the following edge begins. Here are some examples of paths in our sample network:

$$\tt{1\  {\color{red}\leftarrow}\  2 }$$

$$\tt{1\  {\color{red}\leftarrow}\  4\  {\color{red}\leftarrow}\  5 }$$

$$\tt{1\  {\color{red}\leftarrow}\  2\  {\color{red}\leftarrow}\  4 \ {\color{red}\leftarrow}\  2 }$$

$$\tt{4\  {\color{red}\leftarrow}\  3\  {\color{red}\leftarrow}\  1\ {\color{red}\leftarrow}\  4  \ {\color{red}\leftarrow}\  3\  {\color{red}\leftarrow}\  1}$$

The **length** of a path is the number of edges the path consists of. For example, the paths shown above have lengths 1, 2, 3, and 5, respectively. 


## Counting paths

Since a path of length 1 is simply an edge of the network, the adjacency matrix $A = [a_{\tt ij}]$ can be seen as a table which gives, for any nodes $\tt j$ and $\tt i$, the number of paths of length 1 that start at 
$\tt j$  and end at $\tt i$.


It turns out that we can also use the adjacency matrix to calculate the number of paths of other lengths. 
For example, lets say that we want to compute the number of paths of length 2  which start at a node $\tt j$ and end at a node $\tt i$:


$${\tt i}\  {\color{red}\leftarrow} {\ \large \bullet\ } {\color{red}\leftarrow}\  {\tt j}$$

The number of such paths can be 
obtained by multiplying the ${\tt i}^{\text{th}}$ row and the ${\tt j}^{\text{th}}$ column
of the adjacency matrix $A = [a_{\tt ij}]$:

<br/>

$$
\begin{bmatrix}
 a_{\tt i1}& a_{\tt i2} & \cdots & a_{\tt in} \\
\end{bmatrix}
\cdot 
\begin{bmatrix}
a_{\tt 1j}\\
a_{\tt 2j}\\
\vdots \\ 
a_{\tt nj}\\
\end{bmatrix}
\ = \ (a_{\tt i1}\cdot  a_{\tt 1j}) + (a_{\tt i2}\cdot  a_{\tt 2j}) + \dots + (a_{\tt in}\cdot  a_{\tt nj})
$$


Indeed, each product $a_{\tt ik}\cdot  a_{\tt kj}$ is equal to either 0 or 1. Moreover, it is equal to 1 only if $a_{\tt ik} =1$ (which means 
that there is an edge ${\tt i\leftarrow k}$), and $a_{\tt kj} =1$ (which means 
that there is an edge ${\tt k\leftarrow j}$). In other words, $a_{\tt ik}\cdot  a_{\tt kj} = 1$ precisely when we have a length 2 path 

$${\tt i}\ \  {\color{red}\leftarrow} \ \  {\tt k} \ \  {\color{red}\leftarrow} \ \  {\tt j}$$

Since the product of the  ${\tt i}^{\text{th}}$ row and the ${\tt j}^{\text{th}}$ column of the matrix $A$
gives the  entry in the ${\tt i}^{\text{th}}$ row and the ${\tt j}^{\text{th}}$ column of the matrix $A\cdot A$, this proves the following fact:

<br/>

<div class="bg-danger" style="padding:15px; font-size:110%">
<p>
<b>Proposition.</b> If $A = [a_{\tt ij}]$ is the adjacency matrix of a network, then the entry in the ${\tt i}^{\text{th}}$ row and the ${\tt j}^{\text{th}}$ column of the matrix $A^2 = A\cdot A$ is equal to the number of paths of length 2 which which start at the node $\tt j$ and end at the node $\tt i$.
</p>
</div>

<br/>


This can be generalized as follows:

<br/>
<div class="bg-danger" style="padding:15px; font-size:110%">
<p>
<b>Proposition.</b> Let $A = [a_{\tt ij}]$ be the adjacency matrix of a network, and let 
$A^n = A\cdot {\dots}\cdot A$ be the matrix  obtained by multiplying  $A$ by itself $n$ times.
The entry in the ${\tt i}^{\text{th}}$ row and the ${\tt j}^{\text{th}}$ column of  $A^n$ is equal to the number of paths of length $n$ which which start at the node $\tt j$ and end at the node $\tt i$.
</p>
</div>



## Example

We will apply these results to our sample network:

<br/>
<img src="https://raw.githubusercontent.com/bbadzioch/MTH309_F2019/master/notebooks_2024/network.png" style="width: 300px;">

<br/>


First, we load SymPy and enter the adjacency matrix of this network:

In [1]:
from sympy import *
init_printing(use_latex = "mathjax")

A = Matrix([[0, 1, 0, 1, 0], [0, 0, 0, 1, 0], [1, 0, 0, 1, 1], [0, 1, 1, 0, 1], [0, 1, 0, 1, 0]])
A

⎡0  1  0  1  0⎤
⎢             ⎥
⎢0  0  0  1  0⎥
⎢             ⎥
⎢1  0  0  1  1⎥
⎢             ⎥
⎢0  1  1  0  1⎥
⎢             ⎥
⎣0  1  0  1  0⎦

The second power of this matrix gives the number of paths of length 2 between any two nodes:

In [2]:
A**2

⎡0  1  1  1  1⎤
⎢             ⎥
⎢0  1  1  0  1⎥
⎢             ⎥
⎢0  3  1  2  1⎥
⎢             ⎥
⎢1  1  0  3  1⎥
⎢             ⎥
⎣0  1  1  1  1⎦

For example, the entry in the 2$^{\text{nd}}$ column and the 3$^{\text{rd}}$ row tells us that there are 3 paths that start at the node $\tt 2$ and end at the node $\tt 3$. Indeed, these paths are:

$$\tt{3\  {\color{red}\leftarrow}\  1\  {\color{red}\leftarrow}\  2 }$$
$$\tt{3\  {\color{red}\leftarrow}\  4\  {\color{red}\leftarrow}\  2 }$$
$$\tt{3\  {\color{red}\leftarrow}\  5\  {\color{red}\leftarrow}\  2 }$$

The 20$^{\text{th}}$ power of the matrix $A$ gives us the number of paths of length 20 between any two nodes:

In [3]:
A**20

⎡358198  1469586  783942   1715349  1142140⎤
⎢                                          ⎥
⎢245763  1008812  538179   1177170  783942 ⎥
⎢                                          ⎥
⎢573209  2351826  1254575  2745054  1827784⎥
⎢                                          ⎥
⎢538179  2206875  1177170  2576696  1715349⎥
⎢                                          ⎥
⎣358198  1469586  783942   1715349  1142140⎦

We see here, for example, that there are 2,351,826 paths of length 20 which start at the node $\tt 2$ and end at the node $\tt 3$. It would be difficult to explicitly list all these paths!